apex/contrib/csrc/layer_norm/ln_fwd_cuda_kernel.cu

```diff
@@ -0,0 +1,186 @@
+
+  const char *x_ptr = static_cast<const char *>(x_);
+
+  const char *g_ptr = static_cast<const char *>(gamma_);
+  const char *b_ptr = static_cast<const char *>(beta_);
+
+  char *y_ptr = static_cast<char *>(y_);
+  compute_t *mu_ptr = static_cast<compute_t *>(mu_);
+  compute_t *rs_ptr = static_cast<compute_t *>(rsigma_);
+
+  Vec gamma[LDGS];
+  Vec beta[LDGS];
+#pragma unroll
+  for (int it = 0, col = c; it < LDGS; it++) {
+    gamma[it].load_from(g_ptr + col * BYTES_PER_LDG);
+    beta[it].load_from(b_ptr + col * BYTES_PER_LDG);
+    col += THREADS_PER_ROW;
+  }
+
+  constexpr compute_t rn = 1.f / compute_t(Ktraits::COLS);
+  for (int row = r; row < rows; row += gridDim.x * ROWS_PER_CTA) {
+    Vec x[LDGS];
+#pragma unroll
+    for (int it = 0, col = c; it < LDGS; it++) {
+      x[it].load_from(x_ptr + row * BYTES_PER_ROW + col * BYTES_PER_LDG);
+      col += THREADS_PER_ROW;
+    }
+    compute_t xf[LDGS * NUM_ELTS];
+#pragma unroll
+    for (int it = 0; it < LDGS; it++) {
+#pragma unroll
+      for (int jt = 0; jt < NUM_ELTS; jt++) {
+        xf[it * NUM_ELTS + jt] = compute_t(x[it].data.elt[jt]);
+      }
+    }
+
+    compute_t mu_local = 0.f;
+
+#pragma unroll
+    for (int it = 0; it < LDGS; it++) {
+#pragma unroll
+      for (int jt = 0; jt < NUM_ELTS; jt++) {
+        mu_local += xf[it * NUM_ELTS + jt];
+      }
+    }
+
+#pragma unroll
+    for (int it = 1; it < THREADS_PER_WARP; it *= 2) {
+      mu_local += __shfl_xor_sync(uint32_t(-1), mu_local, it);
+    }
+    mu_local *= rn;
+    if(lane == 0){
+    mu_ptr[row] = mu_local;
+    }
+    compute_t var_local = 0.f;
+
+#pragma unroll
+    for (int it = 0; it < LDGS; it++) {
+#pragma unroll
+      for (int jt = 0; jt < NUM_ELTS; jt++) {
+        compute_t diff = xf[it * NUM_ELTS + jt] - mu_local;
+        var_local += diff * diff;
+      }
+    }
+
+#pragma unroll
+    for (int it = 1; it < THREADS_PER_WARP; it *= 2) {
+      var_local += __shfl_xor_sync(uint32_t(-1), var_local, it);
+    }
+    compute_t rsigma = rsqrtf(var_local * rn + epsilon);
+    if(lane == 0){
+    rs_ptr[row] = rsigma;
+    }
+
+#pragma unroll
+    for (int it = 0; it < LDGS; it++) {
+#pragma unroll
+      for (int jt = 0; jt < NUM_ELTS; jt++) {
+        base_t tmp = (rsigma * (xf[it * NUM_ELTS + jt] - mu_local));
+        x[it].data.elt[jt] = gamma[it].data.elt[jt] *  tmp + beta[it].data.elt[jt];
+      }
+    }
+
+#pragma unroll
+    for (int it = 0, col = c; it < LDGS; it++) {
+      x[it].store_to(y_ptr + row * BYTES_PER_ROW + col * BYTES_PER_LDG);
+      col += THREADS_PER_ROW;
+    }
+  }
+}
+template<typename scalar_t>
+void launch(
+    at::Tensor & y, // BxSxhidden_size
+    at::Tensor & mu,
+    at::Tensor & rsigma,
+    const at::Tensor & x, // BxSxhidden_size
+    const at::Tensor & gamma,
+    const at::Tensor & beta,
+    const float epsilon,
+    const int rows,
+    const int cols,
+    const int max_gridx,
+    cudaStream_t stream
+){
+
+  if (cols == 1024) {
```

Comment:
Is there any particular reason to support only tensors with hidden_size 1024?

Reply:
As it's written, this particular kernel supports only dimensions which are multiples of 512B. So BERT base _should_ also work.

Reply:
It seems to me that it only works with dimension of 1024. I did not look deeper than this file and I do not know how kernels are implemented. I tested with different sizes and it raises an exception for any dimension different than 1024:
`
from apex.contrib.layer_norm import FastLayerNorm
dim1=2048
dim2=1024. # 512, 2048, 4096 fail
x = torch.rand(dim1, dim2, dtype=torch.float32).cuda()
fln = FastLayerNorm(dim2).cuda()
y = fln(x)
`

Reply:
Oh yes, I meant that the kernel template can be instantiated with other sizes under the constraint mentioned above.

Reply:
@yjk21 any suggestions on kernel traits for other dimensions (like 2048, 4096, etc)? I tried same as `1024`, but got no perf improvements on `A100`.  I do get better perf with `1024` on A100 as yours